src/apple/macos/ffi.rs

```diff
@@ -38,89 +43,57 @@ extern "C" {
-    //     allocator: CFAllocatorRef,
-    //     options: IOOptionBits,
-    // ) -> kern_return_t;
-    // pub fn IORegistryEntryGetName(entry: io_registry_entry_t, name: *mut c_char) -> kern_return_t;
-}
-
-extern "C" {
-    pub fn CFStringCreateWithCStringNoCopy(
-        alloc: *mut c_void,
-        cStr: *const c_char,
-        encoding: CFStringEncoding,
-        contentsDeallocator: *mut c_void,
-    ) -> CFStringRef;
-
-    // Disk information functions are non-operational on iOS because of the sandboxing
-    // restrictions of apps, so they don't can't filesystem information. This results in
-    // mountedVolumeURLs and similar returning `nil`. Hence, they are MacOS specific here.
-
-    pub fn DASessionCreate(allocator: CFAllocatorRef) -> DASessionRef;
-
-    // pub fn DADiskCreateFromVolumePath(
-    //     allocator: CFAllocatorRef,
-    //     session: DASessionRef,
-    //     path: CFURLRef,
-    // ) -> DADiskRef;
-    pub fn DADiskCreateFromBSDName(
+    pub fn IORegistryEntryCreateCFProperty(
+        entry: io_registry_entry_t,
+        key: CFStringRef,
         allocator: CFAllocatorRef,
-        session: DASessionRef,
-        path: *const c_char,
-    ) -> DADiskRef;
-    // pub fn DADiskGetBSDName(disk: DADiskRef) -> *const c_char;
-
-    pub fn DADiskCopyDescription(disk: DADiskRef) -> CFMutableDictionaryRef;
-}
-
-pub type DADiskRef = *const __DADisk;
-pub type DASessionRef = *const __DASession;
-
-// We need to wrap `DASessionRef` to be sure `System` remains Send+Sync.
-pub struct SessionWrap(pub DASessionRef);
-
-impl Drop for SessionWrap {
-    fn drop(&mut self) {
-        if !self.0.is_null() {
-            unsafe {
-                CFRelease(self.0 as _);
-            }
-        }
-    }
+        options: IOOptionBits,
+    ) -> CFDictionaryRef;
+    pub fn IORegistryEntryGetParentEntry(
+        entry: io_registry_entry_t,
+        plane: io_name_t,
+        parent: *mut io_registry_entry_t,
+    ) -> libc::kern_return_t;
+
+    pub fn IOBSDNameMatching(
+        mainPort: mach_port_t,
+        options: u32,
+        bsdName: *const c_char,
+    ) -> CFMutableDictionaryRef;
+
+    // This is deprecated as of macOS 12.0, but Rust doesn't have a good way to only use the replacement on 12+.
+    pub static kIOMasterPortDefault: mach_port_t;
 }
 
-unsafe impl Send for SessionWrap {}
-unsafe impl Sync for SessionWrap {}
-
 #[cfg(all(
     not(feature = "apple-sandbox"),
     any(target_arch = "x86", target_arch = "x86_64")
 ))]
 mod io_service {
-    use super::mach_port_t;
+    use super::io_object_t;
+    use core_foundation_sys::base::mach_port_t;
+    use libc::{c_char, c_void, size_t};
 
-    #[allow(non_camel_case_types)]
-    pub type io_object_t = mach_port_t;
     #[allow(non_camel_case_types)]
     pub type io_connect_t = io_object_t;
-    #[allow(non_camel_case_types)]
-    pub type io_iterator_t = io_object_t;
+
+    extern "C" {
+        pub fn IOServiceMatching(a: *const c_char) -> *mut c_void;
```

Comment:
```suggestion
        pub fn IOServiceMatching(name: *const cchar) -> CFDictionary;
```

[cchar](https://developer.apple.com/documentation/swift/cchar) is **not** a `c_char` but a `i8`. Please make a type alias for it.

Reply:
Also I think it's `CFMutableDictionaryRef` to be exact.

Reply:
Regarding the `cchar` difference, I think that you might have looked at the wrong set of Apple documentation. `cchar` is a Swift type but we're not using any Swift bindings here. If you look at the [Objective-C docs](https://developer.apple.com/documentation/iokit/1514687-ioservicematching?language=objc) for this function, it lists the type as `const char *name`.

I'll switch in the more-correct return type though 👍

Reply:
Oh you're right. Sorry about that...

Reply:
No worries at all. I've switched this to correctly return a `CFMutableDictionaryRef` as well.